crates/runtime/src/materialize/protocol.rs

```diff
@@ -305,20 +307,42 @@ pub fn send_connector_store(
     let binding_index = meta.binding();
     let binding = &task.bindings[binding_index];
 
-    let key_packed = doc::Extractor::extract_all_owned(&root, &binding.key_extractors, buf);
-    let values_packed = doc::Extractor::extract_all_owned(&root, &binding.value_extractors, buf);
-    let mut doc_json = serde_json::to_string(&binding.ser_policy.on_owned(&root))
-        .expect("document serialization cannot fail");
+    // A note on the order of operations:
+    // The value extractors may contain a special truncation sentinel extractor,
+    // which will write the value of this variable. It's important that we
+    // extract the values last, so that the sentinel can account for truncations
+    // in both the keys and the flow document.
+    let truncation_sentinel = AtomicBool::new(false);
+    let key_packed = doc::Extractor::extract_all_owned(
+        &root,
+        &binding.key_extractors,
+        buf,
+        Some(&truncation_sentinel),
+    );
+
+    let doc_json = if binding.store_document {
+        serde_json::to_string(
+            &binding
+                .ser_policy
+                .on_owned(&root, Some(&truncation_sentinel)),
+        )
+        .expect("document serialization cannot fail")
+    } else {
+        String::new()
+    };
+
+    let values_packed = doc::Extractor::extract_all_owned(
+        &root,
+        &binding.value_extractors,
+        buf,
+        Some(&truncation_sentinel),
+    );
 
     // Accumulate metrics over reads for our transforms.
     let stats = &mut txn.stats.entry(binding_index as u32).or_default();
     stats.2.docs_total += 1;
     stats.2.bytes_total += doc_json.len() as u64;
 
-    if !binding.store_document {
```

Comment:
accidental?

Reply:
This was intentional. `doc_json` is being created above, and I just skip serializing it if it's not needed.